src/constraints/StringConstraints.ts

```diff
@@ -44,3 +47,48 @@ export function stringLengthNe(length: number): IConstraint<string> {
 	const expected = `expected.length !== ${length}`;
 	return stringLengthComparator(ne, 's.string.lengthNe', expected, length);
 }
+
+export function stringRegex(regex: RegExp, type: 'url' | 'uuid' | 'regex' | 'email'): IConstraint<string> {
+	const expected = `expected.${type}`;
+	return {
+		run(input: string) {
+			return regex.test(input) //
+				? Result.ok(input)
+				: Result.err(new ConstraintError(`s.string.${type}`, 'Invalid string format', input, expected));
+		}
+	};
+}
+
+export function stringUrl(options?: UrlOptions): IConstraint<string> {
+	return {
+		run(input: string) {
+			try {
+				const url = new URL(input);
+				if (options?.allowedProtocols && !options.allowedProtocols.includes(url.protocol)) {
+					return Result.err(
+						new ConstraintError(
+							's.string.url',
+							'Invalid URL protocol',
+							input,
+							`expected.url.protocol in ${options.allowedProtocols.join(', ')}`
+						)
+					);
+				}
+				if (options?.allowedDomains && !options.allowedDomains.includes(url.hostname)) {
+					return Result.err(
+						new ConstraintError(
+							's.string.url',
+							'Invalid URL domain',
+							input,
+							`expected.url.domain in ${options.allowedDomains.join(', ')}`
+						)
+					);
+				}
```

Comment:
I'm not a fan of doing those checks over and over, it slows down an already slow operation, not to mention it's also inside a try/catch, which makes performance worse.

Reply:
yeah sure. So should I remove them or keep them as it is?